### DISCUSSION TITLE
Shield feature feedback megathread

### DISCUSSION THREAD

**Discussion by Metadorius** [Phobos-developers/Phobos#253] [Ideas and feedback]
This thread was created to provide a centralized place for your ideas and feedback on how to further improve shields feature. Please note that this is supplementary to our Discord where most of the discussion happens, so I recommend to join Discord anyways to stay in touch. Regardless, the feedback is still welcome.

**Comment by McPwny** [comment]
i know you guys have 0 respect for graphics makers but i will leave the first comment here so it doesnt look like no one has anything to say.

after working to make animations with shields for a while more i have found that the buildup anim delay is a non issue since the 'shield up' section of the animation can be packed into first several frames of the shield and it actually works quite nicely. in terms of animation and regarding buildings only, the biggest issues are that the shield anims cannot accommodate damaged frames which severely limits how shields can be animated. regarding units, simply applying an animation on top of units leaves very few options for animators and i believe that the genuine lack of ways to make it "look good" will undermine it as a feature. additionally, ill mention that having an option to increase the transparency of the shield when the shield is low would be a good move, aesthetically.

now in terms of game functionality, the shield logic is a bit lacking in that the only way to use it is to attach it to something. i suggest implementing some logics that make it more interesting. for example, shields which only turn on when a certain building is present, shields which are applied to objects only when they are a certain house, shields that can be picked up in crates, shields that can be activated by some superweapon, shields that go down when attacking, and so on and so forth.

**Reply by Metadorius** [reply]
0. We have 0 respect for douchebags, not graphic makers.
1. I suppose adding damage states support to shields could work and is a good idea.
2. The ability to dynamically "cast" shields is a widely requested feature indeed, if we touch shields again we'll definitely look into it.

**Reply by McPwny** [reply]
i will also mention quick that it could be cool to have shield anims support remap colors

**Comment by Metadorius** [comment]
@McPwny check the new release for shield manipulation warheads. Animation damages states are still to-do.

**Comment by McPwny** [comment]
okay ive got some new feedback for you. the damage states are working very nicely and it looks like the problem of the animation starting to play before the buildup is complete has been addressed. all in all it is very usable now, but there were a few notable issues in animating it

1: the shield idle animations and shield break animation have a downward y offset of 15 pixels
2: the shield idle animations and shield break animation need negative z adjust to be seen ingame
3: the shield hit animation is drawn from some weird point and doesnt support transparency

the first two problems can be handled in art.ini by someone who knows what they are doing, but fixing the third problem is problematic as it requires manually re-rendering the anim with different X offsets until it aligns properly.

one improvement i can think of is for the hit animation to support damage states since, in my opinion, it looks better to animate the shield part itself to show when hit only than to show all the time as an idle anim.

**Reply by Starkku** [reply]
> 1: the shield idle animations and shield break animation have a downward y offset of 15 pixels
> 2: the shield idle animations and shield break animation need negative z adjust to be seen ingame
> 3: the shield hit animation is drawn from some weird point and doesnt support transparency

1. This is actually a 'problem' with attached animations (i.e attached to object and moves alongside it if possible, such as parachutes, mind control ring, AttachEffect in Ares or in fact, shield idle animations) in general. Save for one special case (weapon `Anim` and its specializations), they are centered at the bottom-most pixel of the object's base cell, which for buildings is the top-most cell. There would be a few solutions for this.

- Simply adjust this in art(md).ini like you already suggested, either via `YDrawOffset` if possible or Phobos-introduced key `UseCenterCoordsIfAttached`.
- Change the default behaviour of attached animations to center correctly on buildings. Not ideal since this will alter outcomes for existing use-cases.
- Make shield animations special case that are correctly centered by default. Would once again require changing existing use-cases (even if to a lesser extent) and potentially introduce confusion as it doesn't work like other attached anims.

2. This is a consequence of not hardcoding these things like vanilla game does (f.ex, mind control ring is hardcoded to -1024 ZAdjust and iron curtain nullification anim to -15). It is probably better to not given that one value might not be suitable for all use cases and someone may want to make a shield that doesn't display over the object for an example. It is maybe not a perfect solution but it gives control to the modder instead of enforcing a specific type of set up.

3. There absolutely nothing that should prevent transparency from working with `HitAnim` and it works just fine in my own mod so unsure of what is going on at your end for that. As for the coordinate, this is basically an inverted issue to 1. as this one is drawn at the actual center of the object the shield is attached to.

Adding a toggle to attach hit anims to the object and possibly damage states isn't a bad idea and might look into implementing those in the future.
